keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,612 @@
+This KEP aims to provide a generic mechanism for the use case, because such use cases are also found for protecting secrets and configmaps, [here](https://github.com/kubernetes/enhancements/pull/2640).
+
+Potential use cases are:
+- Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
+- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
+- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
+- Protect `ConfigMap` while being used by `Pod`
+- Protect `StorageClass` while being used by `PersistentVolume`
+- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+- User defined reference
+
+### Goals
+
+- Provide a generic mechanism that prevent X from being deleted when Y exists
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+
+## Proposal
+
+A new controller to protect in-use objects from being deleted is introduced.
+This controller doesn't have special knowledge on which objects should be protected from deletion for which other objects, instead it gets such dependency information from objects that users or controllers set and protects these objects as defined there.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. It updates objects' dependency information and ask this in-use protection controller to protect the used objects while using objects exist.
+
+#### Story 2
+
+A user sets an object X's dependency information for a particular object Y to be protected by this in-user protection controller. For example, when a user knows that `Service` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a dependency.
+
+### Notes/Constraints/Caveats (Optional)
+
+- Cross namespace dependency isn't allowed by [garbage collector](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents), however, cross namespace dependency will be required for in-use protection controller by the Secret protection use case above,
+- Cyclic dependency can be set by user, which potentially makes dead lock,
+- In some cases, users really wan't to delete the objects that are considered to be protected in general.
+For example, immutable Secrets might need to be deleted while it is used by a Pod to update the value.
+
```

Comment:
From @liggitt  https://github.com/kubernetes/enhancements/pull/2640#issuecomment-887552939

- inverting the relationship direction means that the component responsible for adding/removing the finalizer needs to have perfect visibility to all objects of all types.
- It's also unclear how this avoids races in a quick "create parent, create child, delete parent" scenario